// @oakserver/oak/esm/mod.js
// Copyright 2018-2022 the oak authors. All rights reserved. MIT license.
/**
 * A middleware framework for handling HTTP with Deno.
 *
 * oak works well on both Deno CLI and Deno deploy, and is inspired by
 * [koa](https://koajs.com/). It works well with both the Deno CLI and
 * [Deno Deploy](https://deno.com/deploy).
 *
 * ### Example server
 *
 * A minimal router server which responds with content on `/`. With Deno CLI
 * this will listen on port 8080 and on Deploy, this will simply serve requests
 * received on the application.
 *
 * ```ts
 * import { Application, Router } from "https://deno.land/x/oak/mod.ts";
 *
 * const router = new Router();
 * router.get("/", (ctx) => {
 *   ctx.response.body = `<!DOCTYPE html>
 *     <html>
 *       <head><title>Hello oak!</title><head>
 *       <body>
 *         <h1>Hello oak!</h1>
 *       </body>
 *     </html>
 *   `;
 * });
 *
 * const app = new Application();
 * app.use(router.routes());
 * app.use(router.allowedMethods());
 *
 * app.listen({ port: 8080 });
 * ```
 *
 * ### Using Deno's flash server
 *
 * Currently, Deno's flash server is not the default, even with the `--unstable`
 * flag. In order to use the flash server, you need to provide the
 * {@linkcode FlashServer} to the {@linkcode Application} constructor:
 *
 * ```ts
 * import { Application, FlashServer } from "https://deno.land/x/oak/mod.ts";
 *
 * const app = new Application({ serverConstructor: FlashServer });
 *
 * // register middleware
 *
 * app.listen({ port: 8080 });
 * ```
 *
 * Note the currently Deno's flash server requires the `--unstable` flag. If it
 * isn't present, the application will error on listening.
 *
 * @module
 */
import "./_dnt.polyfills.js";
export { Application } from "./application.js";
export { Context } from "./context.js";
export * as helpers from "./helpers.js";
export { Cookies, } from "./cookies.js";
export * as etag from "./etag.js";
export { HttpRequest } from "./http_request.js";
export { FlashServer, hasFlash } from "./http_server_flash.js";
export { HttpServer as HttpServerNative } from "./http_server_node.js";
export { proxy } from "./middleware/proxy.js";
export { compose as composeMiddleware } from "./middleware.js";
export { FormDataReader } from "./multipart.js";
export { ifRange, MultiPartStream, parseRange } from "./range.js";
export { Request } from "./request.js";
export { REDIRECT_BACK, Response } from "./response.js";
export { Router } from "./router.js";
export { send } from "./send.js";
export { ServerSentEvent } from "./server_sent_event.js";
/** Utilities for making testing oak servers easier. */
export * as testing from "./testing.js";
export { isErrorStatus, isRedirectStatus } from "./util.js";
// Re-exported from `std/http`
export { createHttpError, errors as httpErrors, HttpError, isHttpError, Status, STATUS_TEXT, } from "./deps.js";
