// @oakserver/oak/script/http_request.js
"use strict";
var _HttpRequest_deferred, _HttpRequest_request, _HttpRequest_resolved, _HttpRequest_upgradeWebSocket;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpRequest = void 0;
const tslib_1 = require("tslib");
// Copyright 2018-2022 the oak authors. All rights reserved. MIT license.
const dntShim = tslib_1.__importStar(require("./_dnt.shims.js"));
const maybeUpgradeWebSocket = "upgradeWebSocket" in dntShim.Deno
    ? // deno-lint-ignore no-explicit-any
        dntShim.Deno.upgradeWebSocket.bind(dntShim.Deno)
    : undefined;
/** An abstraction which wraps a {@linkcode Request} from Deno's flash server.
 * The constructor takes a {@linkcode Deferred} which it will resolve when the
 * response is ready.
 *
 * This request can be used in situations where there isn't a specific method
 * to respond with, but where a `Promise<Response>` is accepted as a value. It
 * is specifically designed to work with Deno's flash server.
 */
class HttpRequest {
    constructor(request, deferred, upgradeWebSocket) {
        _HttpRequest_deferred.set(this, void 0);
        _HttpRequest_request.set(this, void 0);
        _HttpRequest_resolved.set(this, false);
        _HttpRequest_upgradeWebSocket.set(this, void 0);
        tslib_1.__classPrivateFieldSet(this, _HttpRequest_deferred, deferred, "f");
        tslib_1.__classPrivateFieldSet(this, _HttpRequest_request, request, "f");
        tslib_1.__classPrivateFieldSet(this, _HttpRequest_upgradeWebSocket, upgradeWebSocket ?? maybeUpgradeWebSocket, "f");
    }
    get remoteAddr() {
        return undefined;
    }
    get headers() {
        return tslib_1.__classPrivateFieldGet(this, _HttpRequest_request, "f").headers;
    }
    get method() {
        return tslib_1.__classPrivateFieldGet(this, _HttpRequest_request, "f").method;
    }
    get url() {
        try {
            const url = new URL(tslib_1.__classPrivateFieldGet(this, _HttpRequest_request, "f").url);
            return tslib_1.__classPrivateFieldGet(this, _HttpRequest_request, "f").url.replace(url.origin, "");
        }
        catch {
            // we don't care about errors, we just want to fall back
        }
        return tslib_1.__classPrivateFieldGet(this, _HttpRequest_request, "f").url;
    }
    // deno-lint-ignore no-explicit-any
    error(reason) {
        if (tslib_1.__classPrivateFieldGet(this, _HttpRequest_resolved, "f")) {
            throw new Error("Request already responded to.");
        }
        tslib_1.__classPrivateFieldGet(this, _HttpRequest_deferred, "f").reject(reason);
        tslib_1.__classPrivateFieldSet(this, _HttpRequest_resolved, true, "f");
    }
    getBody() {
        return {
            body: tslib_1.__classPrivateFieldGet(this, _HttpRequest_request, "f").body,
            readBody: async () => {
                const ab = await tslib_1.__classPrivateFieldGet(this, _HttpRequest_request, "f").arrayBuffer();
                return new Uint8Array(ab);
            },
        };
    }
    respond(response) {
        if (tslib_1.__classPrivateFieldGet(this, _HttpRequest_resolved, "f")) {
            throw new Error("Request already responded to.");
        }
        tslib_1.__classPrivateFieldGet(this, _HttpRequest_deferred, "f").resolve(response);
        tslib_1.__classPrivateFieldSet(this, _HttpRequest_resolved, true, "f");
        return Promise.resolve();
    }
    upgrade(options) {
        if (tslib_1.__classPrivateFieldGet(this, _HttpRequest_resolved, "f")) {
            throw new Error("Request already responded to.");
        }
        if (!tslib_1.__classPrivateFieldGet(this, _HttpRequest_upgradeWebSocket, "f")) {
            throw new TypeError("Upgrading web sockets not supported.");
        }
        const { response, socket } = tslib_1.__classPrivateFieldGet(this, _HttpRequest_upgradeWebSocket, "f").call(this, tslib_1.__classPrivateFieldGet(this, _HttpRequest_request, "f"), options);
        tslib_1.__classPrivateFieldGet(this, _HttpRequest_deferred, "f").resolve(response);
        return socket;
    }
}
exports.HttpRequest = HttpRequest;
_HttpRequest_deferred = new WeakMap(), _HttpRequest_request = new WeakMap(), _HttpRequest_resolved = new WeakMap(), _HttpRequest_upgradeWebSocket = new WeakMap();
